### PR TITLE
Incorrect identically-labeled DataFrame objects Exception comes with df.compare method #50083

### LINKED ISSUE #50083: Incorrect identically-labeled DataFrame objects Exception comes with df.compare method
When we have exact same columns in source and target dataframes and distinct number of rows then we are seeing below exception which according to me is misleading.

"**ValueError: Can only compare identically-labeled DataFrame objects**"

#CODE :

```import pandas as pd
#define DataFrames
df1 = pd.DataFrame({'points': [25, 12, 15, 14],
                   'assists': [5, 7, 13, 12]})

df2 = pd.DataFrame({'points': [25, 12, 15],
                    'assists': [5, 7, 13]})

print(df1)

   points  assists
0      25        5
1      12        7
2      15       13
3      14       12

print(df2)

   points  assists
0      25        5
1      12        7
2      15       13

# Compare both dataframes

df1.compare(df2)

+++++++++++++++++++++++++++++++++ERROR+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/frame.py", line 7117, in compare
    keep_equal=keep_equal,
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/generic.py", line 8424, in compare
    mask = ~((self == other) | (self.isna() & other.isna()))
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/ops/common.py", line 69, in new_method
    return method(self, other)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/arraylike.py", line 32, in __eq__
    return self._cmp_method(other, operator.eq)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/frame.py", line 6851, in _cmp_method
    self, other = ops.align_method_FRAME(self, other, axis, flex=False, level=None)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pandas/core/ops/__init__.py", line 289, in align_method_FRAME
    "Can only compare identically-labeled DataFrame objects"
ValueError: Can only compare identically-labeled DataFrame objects
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

### PR DESCRIPTION
**Pull request by tpackard1**
- [ ] closes #50083
- [ ] All code checks passed

### CODE REVIEW

**Comment by tpackard1** [comment]
First time contributing to the code base and hopefully everything goes smoothly.

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
this is a nice start! could you also:

- fixup the failing tests
- add a whatsnew note

**Comment by tpackard1** [comment]
I will try to fix the failing tests tonight.

**Comment by tpackard1** [comment]
@MarcoGorelli  would the following work as a whatsnew note?
**Whatsnew Note**

- Updated the Exception messages for df.compare method to include that both index and columns that are identically-labeled can be compared.
- Updated the test associated with these Exceptions.

**Comment by MarcoGorelli** [comment]
You can just say that you improved the error message - I think there's some similar entries in the what's new file

No need to mention tests, it's a user facing note

**Comment by tpackard1** [comment]
@MarcoGorelli I have updated the whats new file. Please review whenever you have a moment

File: doc/source/whatsnew/v2.0.0.rst [hunk 1]
```diff
@@ -97,6 +97,7 @@ Other enhancements
 - Added :meth:`Index.infer_objects` analogous to :meth:`Series.infer_objects` (:issue:`50034`)
 - Added ``copy`` parameter to :meth:`Series.infer_objects` and :meth:`DataFrame.infer_objects`, passing ``False`` will avoid making copies for series or columns that are already non-object or where no better dtype can be inferred (:issue:`50096`)
 - :meth:`DataFrame.plot.hist` now recognizes ``xlabel`` and ``ylabel`` arguments (:issue:`49793`)
+- :func:`DataFrame.compare` exception message thrown improved to include that both index and columns that are identically labeled can be compared. (:issue:`50083`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.0.0.rst:100]:
```suggestion
- Improved error message in :func:`DataFrame.compare` to clarify that "identically labelled" refers to both index and columns (:issue:`50083`)
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Looks good, just got a minor comment on the whatsnew note

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Wait, something went wrong when merging/rebasing I presume, you've lost all your changes, all I can see if the whatsnew note change

File: doc/source/whatsnew/v2.0.0.rst (hunk 1 above)
Comment (tpackard1) [reply doc/source/whatsnew/v2.0.0.rst]: I updated the whatsnew note to match your comment above

**Comment by tpackard1** [comment]
I have updated the whatsnew comment and added back in the all the other changes. I wanted to make sure all the checks were passed before requesting a review but some of the checks keep failing with the following error message `ERROR  - SystemError: initialization of _internal failed without raising an exception`. Is this due to the change requested? Also sorry this is taking so much or your time @MarcoGorelli  but I appreciate your time as this process as taught me alot.

**Comment by MarcoGorelli** [comment]
Hey - CI failures are unrelated, they're showing up in other PRs, I'll see what's going on there

File: doc/source/whatsnew/v2.0.0.rst [hunk 2]
```diff
@@ -98,6 +98,7 @@ Other enhancements
 - Added :meth:`Index.infer_objects` analogous to :meth:`Series.infer_objects` (:issue:`50034`)
 - Added ``copy`` parameter to :meth:`Series.infer_objects` and :meth:`DataFrame.infer_objects`, passing ``False`` will avoid making copies for series or columns that are already non-object or where no better dtype can be inferred (:issue:`50096`)
 - :meth:`DataFrame.plot.hist` now recognizes ``xlabel`` and ``ylabel`` arguments (:issue:`49793`)
+- Improved error message in :func:`DataFrame.compare` to clarify that "identically labelled" refers to both index and columns (:issue:`50083`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.0.0.rst:101]: sorry, just noticed that this affects much more than just `compare`, e.g:
```python
In [8]: df == dfother
---------------------------------------------------------------------------
ValueError: Can only compare identically-labeled DataFrame objects
```
would also have thrown this error

let's rephrase as
```python
Improved error message when trying to align :class:`DataFrame` objects (for example, in :func:`DataFrame.compare`) to clarify ...
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
one last minor comment, sorry - then I think it's good to go 🚀

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me, thanks @tpackard1

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @tpackard1

**Pull request merged**